### PR TITLE
UCM/GTEST: cuda memory allocation/free hooks

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6228/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6241/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@yosefe  plz review

**Comment by yosefe** [comment]
@bureddy this PR is too big, need to see how to reduce the amount of code in UCM, maybe with macros. or add only some cuda alloc/free functions in this PR and more in next one,
Also, probably would be better to separate cuda events from event.c to (new file) cuda_event.c

**Comment by bureddy** [comment]
@yosefe  ok. will split it this and move to new file

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6324/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@yosefe added only runtime API in this PR.  will push driver API in separate PR.
Seems moving to separate file need more code organization for lock/unlock and event dispatch. 
will handle it later.

File: src/ucm/event/event.c [hunk 1]
```diff
@@ -396,21 +519,24 @@ static ucs_status_t ucm_event_install(int events)
     }
 
     /* TODO lock */
-    status = ucm_mmap_install(native_events);
-    if (status != UCS_OK) {
-        ucm_debug("failed to install mmap events");
-        goto out_unlock;
-    }
+    if (native_events) {
```
Comment (yosefe) [inline src/ucm/event/event.c]: why need this if?

File: src/ucm/api/ucm.h [hunk 2]
```diff
@@ -37,11 +37,23 @@ typedef enum ucm_event_type {
     UCM_EVENT_VM_MAPPED       = UCS_BIT(16),
     UCM_EVENT_VM_UNMAPPED     = UCS_BIT(17),
 
+    /* Mem type alloc/free events */
```
[the remark this answers is not in the document]
Comment (bureddy) [reply src/ucm/api/ucm.h]: fixed

File: src/ucm/event/event.c [hunk 3]
```diff
@@ -338,21 +338,143 @@ void *ucm_sbrk(intptr_t increment)
...
+
+cudaError_t ucm_cudaFreeHost(void *ptr)
 {
     cudaError_t ret;
+    void *pDevice;
 
     ucm_event_enter();
 
-    ucm_trace("ucm_cudaFree(addr=%p )", addr);
+    ucm_trace("ucm_cudaFreeHost(ptr=%p )", ptr);
```
[the remark this answers is not in the document]
Comment (bureddy) [reply src/ucm/event/event.c]: fixed

File: src/ucm/event/event.c (hunk 1 above)
Comment (bureddy) [reply src/ucm/event/event.c]: removed., not relevant for this PR.  I had this to fix some issue with complete feature PR.  will revisit it in upcoming PRs if this needed,

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6352/ for details (Mellanox internal link).

**Pull request merged**